File: java/src/main/java/example/algorithms/sort/Sort.java
package example.algorithms.sort;

/**
 * Created by Ma Dong on 2018/11/30.
 */
public class Sort {

    public static void selectSort(int[] array){

        for (int i = 0; i < array.length ; i++) {

            int k = i;

            for (int j = i+1; j < array.length ; j++) {

                if (array[k] > array[j]){
                    k = j;
                }
            }

            if ( k != i){
                swap(array,k,i);

            }

        }

    }

    public static void bubleSort(int[] array){

        for (int i = 0; i < array.length; i++) {

            for (int j = 0; j < array.length -1 ; j++) {
                if (array[j] > array[j+1]){
                    swap(array,j,j+1);
                }
            }
        }

    }

    public static void bubleSortPro(int[] array){

        boolean flag = false;
        for (int i = 0; i <array.length ; i++) {

            for (int j = 0; j < array.length -1; j++) {
                if (array[j] > array[j+1]){
                    swap(array,j,j+1);
                    flag = true;
                }
            }
            if (flag){
                flag = false;
            }else {
                break;
            }
        }
    }

    private static void swap(int[] array,int x ,int y){
        array[x] = array[x] ^ array[y];
        array[y] = array[x] ^ array[y];
        array[x] = array[x] ^ array[y];
    }

    public static void main(String[] args) {
        int[] array = {1, 5, 3, 6, 7, 8, 9, 10};
        bubleSort(array);

        for (int i = 0; i <array.length ; i++) {
            System.out.println(array[i]);
        }
    }
}

File: golang/example/algorithms/sort/sort.go
package sort

/**
	选择排序
 */
func SelectSort(arr[] int)  {

	for i:= 0;i< len(arr) ;i++  {
		k := i
		for j := i+1;j< len(arr) ;j++  {
			if arr[k] > arr[j] {
				k = j
			}
		}
		if k != i {
			arr[k],arr[i] = arr[i],arr[k]
		}
	}
}

/**
	冒泡排序
 */
func BubleSort(arr[] int)  {
	for i:=0;i<len(arr) ;i++  {

		for j := 0;j <len(arr) -1 ;j++  {

			if arr[j] > arr[j+1] {
				arr[j],arr[j+1] = arr[j+1],arr[j]
			}
		}

	}
}

/**
	冒泡排序 改进版
 */
func BubleSortPro(arr[] int)  {

	flag := false

	for i:=0;i<len(arr) ;i++  {

		for j := 0;j <len(arr) -1 ;j++  {

			if arr[j] > arr[j+1] {
				arr[j],arr[j+1] = arr[j+1],arr[j]
				flag = true
			}
		}

		if !flag {
			break
		}else {
			flag = false
		}


	}
}
